### ISSUE TITLE
New Rule proposal: Ensure hoisted APIs are placed at top-level

### ISSUE THREAD

**Issue by kirkwaiblinger** [vitest-dev/eslint-plugin-vitest#745]
Since `vi.mock()` is hoisted, it's quite misleading/confusing to read if it's placed inside a conditional or inside a test, and indicates the author probably meant to use `vi.doMock()`. The proposed rule would force `vi.mock()` (and any other statically hoisted API calls if there are any?) to be written at the top level of the module.

### Fail Examples

```ts
if (foo) {
   vi.mock(...);
}
```
```ts

describe('foo', () => {
  it('bar', () => {
    vi.mock(...);
  });
});
```

### Pass Examples

```ts
vi.mock(...);

describe('foo', () => {
  it('bar', () => {
    // ...
  });
});
```

### Further details

I think that two suggestions ought to be provided:

1. Replace `vi.mock()` with `vi.doMock()`
2. Move `vi.mock()` to top level

Looks like `vi.unmock()` and `vi.hoisted()` should probably be treated similarly by this rule.

--- 

I would be happy to send in a PR for this if accepted.

**Comment by kirkwaiblinger** [comment]
Related rule regarding imports: https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/first.md

Could make sense to ensure that hoisted calls are not only at the top level (like `import`/`export` statements are always syntactically required to be) but also before any  runtime code. 

fail:
```ts
console.log('foo');
vi.mock(); // bad; hoisted things should come first
```

pass:
```ts
import foo from './foo.js';
vi.mock(...); 
console.log('bar');
```
